### PR TITLE
Redesign the options page

### PR DESCRIPTION
**Pull request by iamakulov**
I’ve cleaned up the extension options page slightly and improved accessibility. Before and after:

The options page before the cleanup and after it

----

**The list of changes:**

*26b7f33 – Replace spans with labels for better accessibility*
This makes focusing fields and checking checkboxes easier. When you add a label and link it with an input, this label becomes clickable, and you can click it to check the checkbox or focus the input it’s linked with.

*186f087 – Implement proper wrapping*
In forms, line breaks are usually created with `<br/>`. This commit removes the `.caption` width limitation to make wide captions look properly and implements proper line breaks instead.

*9300cbd – Resize checkboxes and align them*
A large checkbox is blurry, a smaller one looks way better. Also, `baseline` alignment is used instead of the default `middle` one because it also looks better.

*c2bdbd2 – Swap labels and checkboxes*
Make the labels appear after the checkboxes to follow common UI patterns. Also, include the captions into the labels to increase the clickable area.

*861e1f7 – Make text input non-bold*
The bold text looks like a design exception and seem to have no meaning. This in fact is just a guess, so please correct me if I’m wrong.

Looking forward to your comments.

### CODE REVIEW

**Comment by iamakulov** [comment]
In fact, the options page requires a wider redesign from the UX point due to the several issues. It isn’t critical, but it’s a positive change that could save each user who tries to configure something several minutes of time. For example, currently it’s unclear that blacklisting/whitelisting of actions will be applied only when the corresponding checkbox is checked, or that “Page urls to inject DevTools in” will be used when the “Inject in all pages” checkbox is unchecked. It _is written_, but it _isn’t recognizable_ from the first sight. A quick-and-dirty solution could be to disable the text boxes when the data from them is not used. A better one involves implementing switches like “Inject into: :radio_button: all pages; :radio_button: specific pages: ________”.

I’ve tried fixing these issues in part of this PR, but the amount of work is quite large. The options page should support radio buttons and should re-render itself when something changes to make applying conditional classes possible. I don’t know if I manage to do this in the near time (most likely no), so I’m just leaving this note here to document (or something) these problems.

**Comment by zalmoxisus** [comment]
Hey @iamakulov,

Thanks a lot for this contribution! I guess those styles problems appear on Windows (due to using other fonts by Chrome or because the scrollbar takes some width), right? I didn't get to test there. Just so you know it looked on MacOS (which uses `Helvetica Neue` by default) like that:
screen

What do you think of making it more compact as above?

**Comment by zalmoxisus** [comment]
About checkboxes for `Filter spec. actions` and `Inject in all pages` totally agree. We should move everything to a React component (instead of rendering directly), add states and hide inactive form elements. It would be clearer and the form would be compact.

**Comment by iamakulov** [comment]
> What do you think of making it more compact as above?

Wow, this form looks way better. I still doubt about the checkbox placement (the common UI pattern is checkbox + label, not label + checkbox), but totally agree that this layout is more suitable.

Yep, let’s try moving options to a separate component and fixing the issues. Will return to this in a week.

**Comment by zalmoxisus** [comment]
Ok. We can move checkboxes to the left side and, as we have enough space there, replace them with [switchers](https://github.com/zalmoxisus/react-switcher).

Instead of adjusting width manually as I did before, se should better use flexbox style with `flex-wrap: nowrap` for every line containing 2 elements.

**Comment by iamakulov** [comment]
I see. Will use switchers and flexbox then.

**Comment by iamakulov** [comment]
@zalmoxisus, a couple of questions while I’m redesigning the options:

* Why is the `autocommit` option a thing? Why are actions automatically committed when they reach like 50 items?
* What does the `State serialization` option do?

**Comment by gaearon** [comment]
@iamakulov 

Just wanted to drop by and thank you for working on this. Most developers including me have troubles figuring out *what exactly* is off about the UI, and help from people like you is dearly appreciated.

**Comment by iamakulov** [comment]
@gaearon, wow, absolutely unexpected! Thank you for kind words.

**Comment by zalmoxisus** [comment]
> Why is the autocommit option a thing? Why are actions automatically committed when they reach like 50 items?
In most of cases we need only the last changes not thousands of logs. It's intended to solve the performance problems for [rendering the monitors](https://github.com/gaearon/redux-devtools-log-monitor/issues/17) and especially for the extension lifecycle.

Chrome doesn't allow the extension to communicate with the app's page directly (though it is declared as being for security reason, it's mostly imposed by design as the page and the extension run in different processes), that's why we relay the changes to the extension's content script, which sends them to the extension's background script, from there the extension's window can use them. This is an expensive operation especially when serializing the states. Though we made a lot of optimizations (now we mostly don't pass the full lifted state, but only the changes) it is still the bottleneck for productivity, and this object shouldn't grow infinitely.

>What does the State serialization option do?
We need this only when [using Immutable.js](https://github.com/zalmoxisus/redux-devtools-extension/issues/35) or [having circular dependencies in states](https://github.com/zalmoxisus/redux-devtools-extension/issues/5). In this cases the states and action's payload objects will be serialized via `jsan`, so we don't lose any data by passing to the extension background script. In other cases we have plain objects and can gain some performance by disabling this option.

**Comment by zalmoxisus** [comment]
@gaearon, exactly, thank you all for all the help here!

**Comment by iamakulov** [comment]
Related: #61.

**Comment by iamakulov** [comment]
Currently busy, will return to this in three-four days. (Just letting you know that I haven’t stopped working here.)

**Comment by zalmoxisus** [comment]
@iamakulov, take your time. When you need to review something you did, let me know.

**Comment by iamakulov** [comment]
Just to let you know: I haven’t forgotten about the PR, don’t have time to finish it yet.

**Comment by zalmoxisus** [comment]
Thanks, @iamakulov, for letting me know, take your time. Probably for now I will just fix the container's width issue for Windows scroll bar, to release `1.0`. Then later we can continue to work on new UI.

**Comment by ericclemmons** [comment]
I'll be introducing React to my WordPress team (O_o) and your project is definitely a selling point.

I'd be interested in helping solve #54 (Inspector Monitor is legit!) once `1.0` goes out, if you'd like.

**Comment by iamakulov** [comment]
@ericclemmons, thank you for the suggestion! I have three things to do to make this PR finished:

- Perform an actual redesign of the dialog. This is more about the UX design than the code and could take several hours.
- Replace checkboxes with `react-switcher`s ([as @zalmoxisus requested](https://github.com/zalmoxisus/redux-devtools-extension/pull/54#issuecomment-183248591)).
- Resolve conflicts with the master branch.

I’d like if you help me with any of these things (either before `1.0` or after it, what’s better for you).

**Comment by zalmoxisus** [comment]
I fixed Windows layout by [increasing the page width](https://github.com/zalmoxisus/redux-devtools-extension/commit/22f6d29e632c328264fd617cdbf2c7c035382c3e), so the scrollbar has enough space now.

Anyway, UI improvement would be welcome. I tried to avoid conflicts with this PR so far. About `react-switcher`, it's not mandatory, you can use checkboxes or other components. I made that component some time ago because didn't found any other switch or checkbox component with inline styles.

**Comment by iamakulov** [comment]
Got some time for this and worked on the UI.

So basically this is the mockup of what I'm going to implement. Sorry for the quality, a pen and a paper have been the only available tools at the moment :-)

### The mockup



Three text pieces in the bottom are marked are with numbers. Here’re they:

1. Recommended. When the number is reached, DevTools start removing the oldest actions. Improves the DevTools performance. [More info](#)
2. Recommended. Required for ImmutableJS state, optional in most other cases. [More info](#)
3. Show the Chrome notifications when errors occur in the app <br> _(In Firefox, “Chrome” is replaced with “Firefox”)_

The _More info_ link is pointed to a more detailed explanation of the option. As a cheap solution, this could be [your comment in this issue](https://github.com/zalmoxisus/redux-devtools-extension/pull/54#issuecomment-188167725): it has a great explanation, but is a part of a large PR which could make understanding it harder. As a more expensive (but better) solution, we could copy your explanation to the repo wiki or anywhere.

### What has changed

* **Options are now grouped**, and the history limit and the serialization options are moved to the end, into the Miscellaneous section. Groups are required for the “Filter actions in DevTools” and “Allow to run” configuration, but to make the UI more consistent, I grouped the remaining options too. 
* The mode of **the action filtering is now changed with radio buttons**. Filtering had three options (don’t filter/hide some/show some), but didn't express this clearly. Radio buttons are a better choice for this. <br> I temporarily removed the note about putting each action from the new line because I didn’t find where to put it, but I’ll return it later.
* **The “Allow to run” configuration is now also changed with radio buttons**. I could’ve left the checkbox, but I prefer the radio buttons in this case as the choice between the options is slightly more obvious. <br> As in action filtering, I temporarily removed the note about RegExps from the new line, but I’ll return it later.
* The miscellaneous **options now have a brief description** (those text pieces marked with numbers). In fact, they had it earlier, but I still found them confusing. I expanded the first two descriptions so that users now have a helpful hints such as “Recommended” and description of _why_ the option is necessary.
* The **“Limit the action history” option** now has a checkbox. As far as I remember, entering 0 is equivalent to disabling this option, but it’s better to express it with a checkbox.

### Changes of the options structure

The old structure of the options object is unsuitable for the new options dialog. Therefore, this update will change it which most likely will result in changes throughout the app. Just to let you know.

To make migration to the new options structure not painful for the users, I’m going to do something like this:

```diff
    <div className="option option_type_checkbox option_value_inject">
-     <input className="option__element" id="inject" type="checkbox" checked={options.inject} onChange={saveOption}/>
+     <input className="option__element" id="aNewProperty" type="checkbox" checked={options.aNewProperty || options.inject} onChange={saveOption}/>
      <label className="option__caption" htmlFor="inject">
        Inject in all pages
        <span className="option__comment">(disable to allow only the urls bellow)</span>
      </label>
    </div>
```

This will save the setting to the new option property, but fall back to the old one when the new one isn’t yet available.

### Timeline
I’m not sure that I’m able to get to code this in the nearest time, and I don’t want to make any promises because you see how it goes. But I still want to finish the PR, and I will continue working on it.

Looking forward to your comments. If you have any questions, I’d be happy to answer.

**Comment by zalmoxisus** [comment]
@iamakulov, that looks great! Thank you for sketching it!

Just a few changes:

- We'll remove selecting monitors from the option page (it will be right in the monitor's window like [here](http://remotedev.io/local/)). So, it will be easier to switch between them, and the option page will not be overwhelmed (as we want also to select monitors for the browser action popup and for the Chrome DevTools panel).

- "Limit the action history" cannot be disabled anymore (no unlimited number of actions stored) as it will slow the extension down on some point (due to https://github.com/zalmoxisus/redux-devtools-extension/pull/72). The user can still put there a very high value, but he will acknowledge that it affects the productivity.

- For "Serialize the state" we want a select with the following values: "No serialization", "Regular serialization", "Deep serialization", see https://github.com/zalmoxisus/redux-devtools-extension/issues/67#issuecomment-191364884 for more details. Or: "disabled", "regular", "deep".

**Comment by iamakulov** [comment]
Cool! I’ll then refresh the sketch when I get to it.

> We'll remove selecting monitors from the option page (it will be right in the monitor's window like here). 

Just ❤️ this solution. It’s much more descriptive and easier to understand than putting the options in the dialog and describing the monitors somewhere else.

File: src/browser/extension/options/options.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import React from 'react';
```
Comment (iamakulov) [inline src/browser/extension/options/options.js]: As we discussed, moved the options component into a separate file. (Probably it’s better to capitalize the file name, but I can’t change this as I’m using Windows with case-insensitive file system.)

File: src/browser/extension/options/options.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+import FilterGroup from './FilterGroup';
+import AllowToRunGroup from './AllowToRunGroup';
+import MiscellaneousGroup from './MiscellaneousGroup';
+
+export default (props) => (
+  <div>
+    <FilterGroup {...props} />
```
Comment (iamakulov) [inline src/browser/extension/options/options.js]: Split the `Options` component into three subcompontents and moved them to separate files because the component got too long.

File: src/browser/extension/options/syncOptions.js [hunk 3]
```diff
@@ -1,39 +1,64 @@
...
     }, function(items) {
-      options = items;
-      callback(items);
+      options = migrateOldOptions(items);
+      callback(options);
     });
   }
 };
 
+const subscribe = callback => {
```
Comment (iamakulov) [inline src/browser/extension/options/syncOptions.js]: Added options subscription. This is used to re-render the `Options` component when something gets changed.

File: src/browser/extension/options/syncOptions.js [hunk 4]
```diff
@@ -62,6 +87,7 @@ export const isAllowed = (localOptions = options) => (
 export default function syncOptions(toAllTabs) {
   return {
     save: save(toAllTabs),
-    get: get
+    get: get,
+    subscribe: subscribe
```
Comment (iamakulov) [inline src/browser/extension/options/syncOptions.js]: Added options subscription. This is used to re-render the `Options` component when something gets changed.

File: src/browser/extension/options/syncOptions.js [hunk 5]
```diff
@@ -1,39 +1,64 @@
...
       whitelist: '',
       blacklist: '',
       serialize: true,
       notifyErrors: true,
       inject: true,
       urls: '^https?://localhost|0\\.0\\.0\\.0:\\d+\n^https?://.+\\.github\\.io'
     }, function(items) {
-      options = items;
-      callback(items);
+      options = migrateOldOptions(items);
```
Comment (iamakulov) [inline src/browser/extension/options/syncOptions.js]: In the new design, the `filter` option isn’t Boolean anymore and can have three values: “Don’t filter”, “Filter whitelisted” and “Filter blacklisted”. The old `options` structure was inappropriate for this, so I added a method to migrate the old options to the new format. This makes the update painless for the users.

File: src/browser/extension/utils/filters.js [hunk 6]
```diff
@@ -1,5 +1,11 @@
 import mapValues from 'lodash/mapValues';
 
+export const FilterState = {
```
Comment (iamakulov) [inline src/browser/extension/utils/filters.js]: The filter states are used in different parts of the code, so I made an exported enum with them.

File: src/browser/extension/options/index.js [hunk 7]
```diff
@@ -1,79 +1,23 @@
 import React from 'react';
 import { render } from 'react-dom';
+import Options from './options';
 
 chrome.runtime.getBackgroundPage(background => {
   const syncOptions = background.syncOptions;
 
-  const saveOption = e => {
```
Comment (iamakulov) [inline src/browser/extension/options/index.js]: I moved the `saveOption` logics from `index.js` to the concrete place where the options are saved (see [`MiscellaneousGroup.js:54`](https://github.com/zalmoxisus/redux-devtools-extension/pull/54/files#diff-3c2ffc337e1475e7b6bb59de0048a94bR18) for example). I believe it’s simpler and more flexible to let the options know how to store the data instead of writing a generic method that will guess this based on the control type.

File: src/browser/extension/options/index.js [hunk 8]
```diff
@@ -1,79 +1,23 @@
...
-    else if (
-      e.target.type === 'input' || e.target.type === 'text'
-    ) value = Number(e.target.value);
-    else value = e.target.value.trim();
-    syncOptions.save(e.target.id, value);
+  const saveOption = (name, value) => {
+    syncOptions.save(name, value);
   };
 
-  const isValidRegex = str => {
```
Comment (iamakulov) [inline src/browser/extension/options/index.js]: I also removed the `isValidRegex` and `saveUrls` functions as a part of [moving `saveOption`](https://github.com/zalmoxisus/redux-devtools-extension/pull/54/files#r70189641). They don’t actually do anything specific, and removing them helped me to move `saveOption` logics to the new place.

File: src/browser/extension/options/MiscellaneousGroup.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
+import React from 'react';
+
+export default ({ options, saveOption }) => {
+  const browserName = typeof chrome === 'undefined' ? 'Firefox' : 'Chrome';
```
Comment (iamakulov) [inline src/browser/extension/options/MiscellaneousGroup.js]: Is this check sufficient?

File: src/browser/extension/options/MiscellaneousGroup.js [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    <fieldset className="option-group">
+      <legend className="option-group__title">Miscellaneous</legend>
+
+      <div className="option option_value_max-age">
+        <label className="option__label" htmlFor="maxAge">Limit the action history to</label>
+        {' '}
+        <input className="option__element"
+               id="maxAge"
+               type="number"
+               min="1"
```
Comment (iamakulov) [inline src/browser/extension/options/MiscellaneousGroup.js]: I made the `maxAge` input a `number` input (so it has nice buttons to change the value with mouse) and limited it down to 1. Is the limit OK?

File: src/browser/extension/inject/index.js [hunk 11]
```diff
@@ -5,7 +5,7 @@ window.devToolsExtensionID = 'lmhkpmbekcpmknklioeibfkpmmfibljd';
 
 chrome.runtime.sendMessage(window.devToolsExtensionID, { type: 'GET_OPTIONS' }, function(response) {
   if (!response.options.inject) {
-    const urls = response.options.urls.split('\n').join('|');
+    const urls = response.options.urls.split('\n').filter(Boolean).join('|');
```
Comment (iamakulov) [inline src/browser/extension/inject/index.js]: This thing is a bit weird, but it’s a quick solution.

So the thing is that previously we saved and synchronized the trimmed content of the `blacklist`, `whitelist` and `url` textareas. When a user entered

```
AN_ACTION
ANOTHER_ACTION
(a new line)
```

into the `blacklist` field, we trimmed it and received `"AN_ACTION\nANOTHER_ACTION"` string. Then we replaced `\n` with `|` and used that string as a regexp. The trimming was necessary because if we wouldn’t do it, we’d receive a `AN_ACTION|ANOTHER_ACTION|` regexp that matched everything. This worked great.

When I moved `Options` into a separate component and made all the inputs controlled, turned out re-rendering an input with the trimmed text effectively prevents a user from putting a new line. The quickest solution for this was to save the non-trimmed version of the text and instead filter out all the empty lines where a regexp is created. `.filter(Boolean) does exactly this: it filters out all falsy values including empty strings.

There’s probably a better solution for this, but I haven’t investigated.

File: src/browser/views/options.jade [hunk 12]
```diff
@@ -9,37 +9,85 @@ html
...
+    .option_type_checkbox .option__hint,
+    .option_type_radio .option__textarea,
+    .option_type_radio .option__hint {
+      margin-left: 20px;
+    }
+
+
+    /* Checkbox styling */
+    .option_type_checkbox .option__element {
+      /* Radio button in Chrome is 2px narrower than the checkbox.
```
Comment (iamakulov) [inline src/browser/views/options.jade]: A doubtful moment. Here I assume that the only browser we support is Chrome. I should’ve probably try to build and install the Firefox extension and check how the checkboxes look there, but I haven’t found any documentation on this so I went to the next task. Should we test this in Firefox?

**Comment by iamakulov** [comment]
Apart from those comments, there’re a couple of more general ones:

* We had a stub function for showing errors when a user enters an invalid regexp, but it didn’t do anything, and I removed it. In fact, we _need_ to show an error in such cases, but let’s do this as a separate task so that this PR isn’t blocked.
* [You’ve told](https://github.com/zalmoxisus/redux-devtools-extension/pull/54#issuecomment-216465854) that the “Serialize state” option should be a select box, but as far as I understand the new serialization values aren’t implemented yet so I didn’t design and implement the switcher for them. Am I right?

**Comment by zalmoxisus** [comment]
Looks amazing. Thanks a lot for implementing this! i'll respond inline.

> We had a stub function for showing errors when a user enters an invalid regexp, but it didn’t do anything, and I removed it. In fact, we need to show an error in such cases, but let’s do this as a separate task so that this PR isn’t blocked.

OK.

> You’ve told that the “Serialize state” option should be a select box, but as far as I understand the new serialization values aren’t implemented yet so I didn’t design and implement the switcher for them. Am I right?

We can start it from here as it would be a part of `migrateOldOptions` function. Having `1` and `2` instead of `true` will not break the current implementation. I will implement the last option after this PR, as [it's just commented there](https://github.com/zalmoxisus/redux-devtools-extension/blob/master/src/browser/extension/utils/contentScriptMsg.js#L6-L11).

> It’d be great to write the option descriptions for the “More info” links in Wiki (as I mentioned in #54 (comment)). Can we do this? I can help with writing texts. (I think we’re better to do this as a separate task to not block this PR.)

Agree about making it in a separate PR. Maybe better to add it in [docs](https://github.com/zalmoxisus/redux-devtools-extension/tree/master/docs), so we could later move it to GitBook?

File: src/browser/extension/options/options.js (hunk 1 above)
Comment (zalmoxisus) [inline src/browser/extension/options/options.js]: Not sure it works on Windows, but my MacOS is also case insensitive and `git mv` to a random name then `git mv` back with upper case worked for me in the past. I can do it after merging the PR.

File: src/browser/extension/options/MiscellaneousGroup.js [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+               value={options.maxAge}
+               onChange={(e) => saveOption('maxAge', Number(e.target.value))}/>
+        {' '}
+        <label className="option__label" htmlFor="maxAge">items</label>
+        <div className="option__hint">
+          When the number is reached, DevTools starts removing the oldest actions. Improves the DevTools performance.
```
Comment (iamakulov) [inline src/browser/extension/options/MiscellaneousGroup.js]: Not sure about plurality here. What sounds better to you: “DevTools starts” or “DevTools start”?

File: src/browser/extension/options/MiscellaneousGroup.js [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+          {' '}<a href="https://github.com/zalmoxisus/redux-devtools-extension/pull/54#issuecomment-188167725">More info</a>
```
Comment (iamakulov) [inline src/browser/extension/options/MiscellaneousGroup.js]: It’s probably better to write this in two lines:

```js
{' '}
<a href="https://github.com/zalmoxisus/redux-devtools-extension/pull/54#issuecomment-188167725">More info</a>
```

Will do this in the evening.

File: src/browser/extension/options/MiscellaneousGroup.js (hunk 9 above)
Comment (zalmoxisus) [inline src/browser/extension/options/MiscellaneousGroup.js]: Firefox implemented WebExtension API, which is chrome compatible, so `chrome` object is present there. So probably we should use `navigator.userAgent.indexOf("Firefox") === -1`.

File: src/browser/extension/options/MiscellaneousGroup.js (hunk 10 above)
Comment (zalmoxisus) [inline src/browser/extension/options/MiscellaneousGroup.js]: [`1` throws an exception](https://github.com/gaearon/redux-devtools/pull/267), it should be at least `2`.

File: src/browser/views/options.jade (hunk 12 above)
Comment (zalmoxisus) [inline src/browser/views/options.jade]: No. We'll do it as a separate PR when we'll be ready for that. It will look different in FF for sure, as we're including [the Chrome css file here](https://github.com/zalmoxisus/redux-devtools-extension/blob/master/src/browser/extension/manifest.json#L49).

Comment (iamakulov) [inline src/browser/views/options.jade]: BTW, the comment is outdated, will update it today. (Adding `#todo` for this discussion to search later.)

File: src/browser/extension/options/MiscellaneousGroup.js (hunk 10 above)
Comment (iamakulov) [inline src/browser/extension/options/MiscellaneousGroup.js]: `#todo`

File: src/browser/extension/options/MiscellaneousGroup.js (hunk 9 above)
Comment (iamakulov) [inline src/browser/extension/options/MiscellaneousGroup.js]: `#todo`

File: src/browser/extension/options/options.js (hunk 1 above)
Comment (iamakulov) [inline src/browser/extension/options/options.js]: Will try to this myself.

`#todo`

File: src/browser/extension/options/MiscellaneousGroup.js (hunk 13 above)
Comment (zalmoxisus) [inline src/browser/extension/options/MiscellaneousGroup.js]: Maybe `When the number is reached, the oldest actions are getting removed.` :-)

Comment (iamakulov) [inline src/browser/extension/options/MiscellaneousGroup.js]: I usually prefer [active voice](http://examples.yourdictionary.com/examples-of-active-and-passive-voice.html) over the passive one because it’s more explicit, and people perceive it easier. So I’d better stick with `DevTools start(s)`. :–)

Comment (zalmoxisus) [inline src/browser/extension/options/MiscellaneousGroup.js]: It would be confusing to indicate that it could be one or more. Actually, [`redux-devtools-instrument`](https://github.com/zalmoxisus/redux-devtools-instrument) is removing, si better to use "the instrumentation enhancer starts removing", or just "the enhancer". It will also indicate that it is referred only to the case when using the extension as a Redux enhancer. As the extension can be used now [not only for Redux](https://github.com/zalmoxisus/redux-devtools-extension/blob/master/docs/API/Methods.md#windowdevtoolsextensionconnectconfig).

Comment (iamakulov) [inline src/browser/extension/options/MiscellaneousGroup.js]: > It would be confusing to indicate that it could be one or more

Huh, I’ve actually meant not that we should write exactly “DevTools start(s)”, but that we should write either “DevTools start” or “DevTools starts” :–)

As to “enhancer” instead of “DevTools”, I wouldn’t do this. “Enhancer” is more complex than “DevTools” because it’s an implementation detail, and the users may not know it.

Comment (iamakulov) [inline src/browser/extension/options/MiscellaneousGroup.js]: Will change this to “DevTools start”.

**Comment by iamakulov** [comment]
> We can start it from here as it would be a part of `migrateOldOptions` function.

What if I do this in a separate PR too? This way, we can release the new options dialog earlier.  What do you think?

> Agree about making it in a separate PR. Maybe better to add it in docs, so we could later move it to GitBook?

Sure! Then my duty is to write and submit the text. `#todo`

**Comment by zalmoxisus** [comment]
> What if I do this in a separate PR too?

OK, we can do it after this PR.

File: src/browser/extension/options/MiscellaneousGroup.js (hunk 14 above)
Comment (iamakulov) [inline src/browser/extension/options/MiscellaneousGroup.js]: Done.

File: src/browser/extension/options/MiscellaneousGroup.js (hunk 13 above)
Comment (iamakulov) [inline src/browser/extension/options/MiscellaneousGroup.js]: Done.

Comment (iamakulov) [inline src/browser/extension/options/MiscellaneousGroup.js]: Done.

File: src/browser/views/options.jade (hunk 12 above)
Comment (iamakulov) [inline src/browser/views/options.jade]: Done.

File: src/browser/extension/options/MiscellaneousGroup.js (hunk 10 above)
Comment (iamakulov) [inline src/browser/extension/options/MiscellaneousGroup.js]: Done.

File: src/browser/extension/options/MiscellaneousGroup.js (hunk 9 above)
Comment (iamakulov) [inline src/browser/extension/options/MiscellaneousGroup.js]: Done.

File: src/browser/extension/options/options.js (hunk 1 above)
Comment (iamakulov) [inline src/browser/extension/options/options.js]: Done.

**Comment by iamakulov** [comment]
@zalmoxisus, I think I’m done here. Is there something we should yet do before merging?

The Travis build is failed for some reason, however, both AppVeyor and my local `npm run test` are OK.

**Comment by zalmoxisus** [comment]
@iamakulov, thanks a lot for all the changes! It looks good to me. The only problem I see is that if `Limit the action history` value is less than `2`, it will broke instrumentation. The user will get an error in the console `DevTools.instrument({ maxAge }) option, if specified, may not be less than 2.`, which is not obvious from where it comes. The notification you added that it should greater or equal to 2 could help, but maybe better to forbid setting it to a lower value?

BTW, it looks great also on Firefox:
screen shot 2016-07-18 at 10 55 59 am

**Comment by iamakulov** [comment]
> The notification you added that it should greater or equal to 2 could help. I think better to forbid setting it to a lower value, though we can do it later.

Hm, do I properly understand that you suggest to add a text note like “Can’t be less than 2”?
As to the limit, the value [is already limited to 2](https://github.com/zalmoxisus/redux-devtools-extension/pull/54/files#diff-3c2ffc337e1475e7b6bb59de0048a94bR16).

> BTW, it looks great also on Firefox:

Hm, I dislike the font, and some items are shifted a bit. I think we can do it better.
You’ve told something about providing a Firefox-specific css file. Can I ask you to create it in the place you find appropriate (could be in a separate branch)? I’ll submit a PR with the changes I believe are necessary.

As to other PRs I’ve promised to make, I’m going to submit them soon.

**Comment by zalmoxisus** [comment]
>Hm, do I properly understand that you suggest to add a text note like “Can’t be less than 2”?
As to the limit, the value is already limited to 2.

I tried to set `1` and it works, though it is breaking the user's application. So we shouldn't allow setting this value. We can also make the input's border red to show that there's an error.

> Hm, I dislike the font, and some items are shifted a bit. I think we can do it better.
You’ve told something about providing a Firefox-specific css file. Can I ask you to create it in the place you find appropriate (could be in a separate branch)? I’ll submit a PR with the changes I believe are necessary.

Chrome is injecting its css file, but firefox [not](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/manifest.json/options_ui), unfortunately. To deploy it, you need Firefox Nightly or Developer Edition (as Firefox 48 is not released yet). Then just run `npm run build:firefox` and [load the extension's folder](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Temporary_Installation_in_Firefox) `./build/firefox`. After that yo can just as usually click `Settings` button from our popup (page action page) to open the options page.

**Comment by zalmoxisus** [comment]
Hey @iamakulov,

Just to let you know that I've implemented deep serialization in https://github.com/zalmoxisus/redux-devtools-extension/commit/8754f4a4bbd38de6d08e0408af3a90cd71066724. There's no need for `migrateOldOptions`, `true`/`false` works as well as `0`/`1`/`2`.

**Comment by iamakulov** [comment]
@zalmoxisus, cool, thanks for the notification!

**Comment by born2net** [comment]
I am using the plugin with a large data set (about 700 items) and when I hit about 1000, things start really slowing down in the application. If I disable the plugin I can go to 20K+ with no performance lag.
I have already disabled the history (down to 5).
I am using it with ngrx not sure if this could be the cause?

tx as always,

Sean.

**Comment by zalmoxisus** [comment]
@born2net, this options are used by Redux only. For `ngrx` you have to reimplement them from `ngrx/store-devtools`. Or you could just pass it as parameter ([as you noted](https://github.com/zalmoxisus/redux-devtools-extension/issues/282#issuecomment-272077163) already) instead of using the global variable. Another good feature we have here, is that you can pause the recording, so there will be no perf impact at all, but again it should be implemented in `ngrx/store-devtools`.

**Comment by born2net** [comment]
thank you, I was able to fix the issue using the maxAge
```
StoreDevtoolsModule.instrumentStore({maxAge: 2}),
```
Angular 2 Kitchen sink: http://ng2.javascriptninja.io
and source@ https://github.com/born2net/Angular-kitchen-sink
Regards,

Sean
